tests/codex/blockexchange/engine/testengine.nim

```diff
@@ -344,9 +349,168 @@ asyncchecksuite "NetworkStore engine handlers":
+
+    seckey = PrivateKey.random(rng[]).tryGet()
+    peerId = PeerId.init(seckey.getPublicKey().tryGet()).tryGet()
+    wallet = WalletRef.example
+    blockDiscovery = Discovery.new()
+    peerStore = PeerCtxStore.new()
+    pendingBlocks = PendingBlocksManager.new()
+
+    localStore = CacheStore.new()
+    network = BlockExcNetwork()
+
+    discovery =
+      DiscoveryEngine.new(localStore, peerStore, network, blockDiscovery, pendingBlocks)
+
+    advertiser = Advertiser.new(localStore, blockDiscovery)
+
+    engine = BlockExcEngine.new(
+      localStore, wallet, network, discovery, advertiser, peerStore, pendingBlocks
+    )
+
+    peerCtx = BlockExcPeerCtx(id: peerId)
+    engine.peers.add(peerCtx)
+
+  test "Should exhaust retries":
+    var
+      retries = 2
+      address = BlockAddress.init(blocks[0].cid)
+
+    proc sendWantList(
+        id: PeerId,
+        addresses: seq[BlockAddress],
+        priority: int32 = 0,
+        cancel: bool = false,
+        wantType: WantType = WantType.WantHave,
+        full: bool = false,
+        sendDontHave: bool = false,
+    ) {.gcsafe, async.} =
+      check wantType == WantHave
+      check not engine.pendingBlocks.isInFlight(address)
+      check engine.pendingBlocks.retries(address) == retries
+      retries -= 1
+
+    engine.pendingBlocks.blockRetries = 2
+    engine.pendingBlocks.retryInterval = 10.millis
+    engine.network =
+      BlockExcNetwork(request: BlockExcRequest(sendWantList: sendWantList))
+
+    let pending = engine.requestBlock(address)
+
+    expect RetriesExhaustedError:
+      discard (await pending).tryGet()
+
+  test "Should retry block request":
+    var
+      address = BlockAddress.init(blocks[0].cid)
+      steps = newAsyncEvent()
+
+    proc sendWantList(
+        id: PeerId,
+        addresses: seq[BlockAddress],
+        priority: int32 = 0,
+        cancel: bool = false,
+        wantType: WantType = WantType.WantHave,
+        full: bool = false,
+        sendDontHave: bool = false,
+    ) {.gcsafe, async.} =
+      case wantType
+      of WantHave:
+        check engine.pendingBlocks.isInFlight(address) == false
+        check engine.pendingBlocks.retriesExhausted(address) == false
+        steps.fire()
+      of WantBlock:
+        check engine.pendingBlocks.isInFlight(address) == true
+        check engine.pendingBlocks.retriesExhausted(address) == false
+        steps.fire()
+
+    engine.pendingBlocks.blockRetries = 10
+    engine.pendingBlocks.retryInterval = 10.millis
+    engine.network = BlockExcNetwork(
+      request: BlockExcRequest(
+        sendWantList: sendWantList, sendWantCancellations: NopSendWantCancellationsProc
+      )
+    )
+
+    let pending = engine.requestBlock(address)
+    await steps.wait()
+
+    # add blocks precense
+    peerCtx.blocks = blocks.mapIt(
+      (it.address, Presence(address: it.address, have: true, price: UInt256.example))
+    ).toTable
+
+    steps.clear()
+    await steps.wait()
```

Comment:
Comments?

\# waiting for first "WantBlock" being sent after receiving block presence message
\# `WantBlock` will trigger `blockDeliveryHandler` on the remote peer